source/unified-test-format/unified-test-format.rst

```diff
@@ -0,0 +1,1896 @@
+===================
+Unified Test Format
+===================
+
+:Spec Title: Unified Test Format
+:Spec Version: 1.0.0
+:Author: Jeremy Mikola
+:Advisors: Prashant Mital, Isabel Atkinson
+:Status: Draft
+:Type: Standards
+:Minimum Server Version: N/A
+:Last Modified: 2020-08-21
+
+.. contents::
+
+--------
+
+Abstract
+========
+
+This project defines a unified schema for YAML and JSON specification tests,
+which run operations against a MongoDB deployment. By conforming various spec
+tests to a single schema, drivers can implement a single test runner to execute
+acceptance tests for multiple specifications, thereby reducing maintenance of
+existing specs and implementation time for new specifications.
+
+META
+====
+
+The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in `RFC 2119 <https://www.ietf.org/rfc/rfc2119.txt>`__.
+
+Specification
+=============
+
+
+Terms
+-----
+
+Entity
+  Any object or value that is indexed by a unique name and stored in the
+  `Entity Map`_. This will typically be a driver object (e.g. client, session)
+  defined in `createEntities`_ but may also be a
+  `saved operation result <operation_saveResultAsEntity_>`_. Entities are
+  referenced throughout the test file (e.g. `Entity Test Operations`_).
+
+Internal MongoClient
+  A MongoClient created specifically for use with internal test operations, such
+  as inserting collection data before a test, configuring fail points during a
+  test, or asserting collection data after a test.
+
+
+Schema Version
+--------------
+
+This specification and the `Test Format`_ follow
+`semantic versioning <https://semver.org/>`__. Backwards breaking changes (e.g.
+removing a field, introducing a required field) will warrant a new major
+version. Backwards compatible changes (e.g. introducing an optional field) will
+warrant a new minor version. Small bug fixes and internal changes (e.g. grammar)
+will warrant a new patch version.
+
+Each test file defines a `schemaVersion`_, which test runners will use to
+determine compatibility (i.e. whether and how the test file will be
+interpreted). Test runners MAY support multiple versions of the test format.
+Test runners MUST NOT process incompatible files but are otherwise free to
```

Comment:
Are you referring to "Test runners MAY support multiple schema versions" or "MUST NOT process incompatible files but MAY determine how to handle such files (e.g. skip, fail)."

Reply:
I'm referring to "MAY determine how to handle such files (e.g. skip, fail)." At least in Java driver we would not easily notice new skipped tests as they don't fail the build, and there are always numerous skipped tests based on server version, topology type, etc.

Reply:
I think it would be helpful to prescribe what should happen when an unknown operation is encountered. I suggest erroring  in this case.

If this is not agreeable I suggest adding a rationale why the different behaviors are acceptable, i.e. why one language runner can error and another can pass the same test file when the file contains an unsupported operation.

Reply:
In that case, I'd rather say that drivers SHOULD fail the test. I'm trying not to be prescriptive here because drivers have their own test libraries. In PHPLIB, I was considering marking these tests incomplete rather than failed, which is a compromise between skipping and failing.

Reply:
> I think it would be helpful to prescribe what should happen when an unknown operation is encountered. I suggest erroring in this case.

This is specified in "Executing an Operation" (test runner MUST raise an error), and I recently added the same for unknown/unsupported keys in `operation.arguments` per https://github.com/mongodb/specifications/pull/846#discussion_r475877714

Reply:
@jyemin: Note that `schemaVersion` is what tells us to even look for individual tests in the `tests` array field (even if that's unlikely to ever change). If a test runner needs to skip/fail tests due to `schemaVersion`, I expect it would want to do so for each element of `tests` within that file (and also using the top-level and test-level `description` fields to derive the name of each test). For example, an unsupported file with five elements in `tests` yields five failures for the test suite.

This seems like a chicken and egg problem, at least for some test libraries. In PHPLIB, which uses PHPUnit, we feed a data provider with a list of JSON files, which are all parsed to yield individual tests (see: [here](https://github.com/mongodb/mongo-php-library/blob/1.7.0/tests/SpecTests/CrudSpecTest.php#L96)).

Reply:
The incomplete status is a good point, how about  saying something like

> The driver SHOULD fail the test. In languages where there are additional test outcomes like Pending/Incomplete, which are distinct from Failed/Skipped, the driver MAY use those outcomes to indicate a test uses unsupported format.

I haven't gotten to operation execution yet but will look at that later.

Reply:
@jmikola can you summarize where we are not on this thread, so that it can be resolved?

Reply:
I think this has been resolved now.